deepchem/models/keras_model.py

```diff
@@ -969,14 +971,89 @@ def restore(self, checkpoint=None, model_dir=None):
 
   def get_global_step(self):
     """Get the number of steps of fitting that have been performed."""
     if tf.executing_eagerly():
       return int(self._global_step)
     return self._global_step.eval(session=self.session)
 
+  def default_assignment_map(self, source_model, include_top=True, **kwargs):
+    """
+    Creates a default assignment map between layers of source and current model.
+    This is used only when a custom assignment map is missing.
+
+    Parameters
+    ----------
+    source_model: dc.models.KerasModel
+        Source model to copy variable values from.
+    include_top: bool, default True
+        if true, copies the last dense layer
+    """
+    assignment_map = {}
+
+    for idx, layer in enumerate(source_model.model.layers[:-1]):
+      assignment_map[layer] = self.model.layers[idx]
+    if include_top:
+      assignment_map[source_model.model.layers[-1]] = self.model.layers[-1]
+
+    return assignment_map
+
+  def load_from_pretrained(self,
+                           source_model,
+                           assignment_map=None,
+                           checkpoint=None,
+                           model_dir=None,
+                           include_top=True,
+                           **kwargs):
+    """Loads a set of layer weights from a pretrained model. The method takes in
+    a source model and an assignment map, and several other optional arguments.
+    The assignment map is a dictionary between the layers of the source model
+    and the layers in the current model, whose weights we want to copy over. If
+    an assignment map is not supplied, a default assignment map is generated,
+    which assumes the architecture has a Dense layer as its last layer and can be
+    included by setting include_top to True. The default assignment map allows
+    applying an existing trained model to a different multi-task setting involving
+    the same model or to a different task like classification from regression or
+    vice-versa.
+
+    Parameters
+    ----------
+    assignment_map: Dict, default None
+      Dictionary containing layer mapping between source and current model layers
+    checkpoint: str, default None
+      the path to the checkpoint file to load.  If this is None, the most recent
+      checkpoint will be chosen automatically.  Call get_checkpoints() to get a
+      list of all available checkpoints.
+    model_dir: str, default None
+      Restore model from custom model directory if needed
+    include_top: bool, default True
+        if true, copies the last dense layer. Used only when assignment map is None
+    """
+    self._ensure_built()
+    if assignment_map is None:
+      assignment_map = self.default_assignment_map(
+          source_model=source_model, include_top=include_top)
+
+    self._assign_ops = []
+    if tf.executing_eagerly():
+      source_model.restore(model_dir=model_dir, checkpoint=checkpoint)
```

Comment:
I'm a bit nervous about calling `restore()` here.  `source_model` is just supposed to be the place values are taken from.  You don't expect it to get modified by this call.  This could lead to a lot of surprises for users.

Reply:
`source_model` is just a container to copy values into, like a dummy and the idea was to have a checkpoint to restore from. But I could add a toggle that allows users to restore if they want to, else just use the current values of a model trained in the same script as finetuning.

Reply:
I realize that `restore` needs to be the there because the pretrained and fine-tuned model run under different Sessions, and the value of the same variables are different between sessions. `tf.assign` is not capable of copying values from one session to another. In this case, values need to be written to disk and hence `restore` is needed. I will mention this explicitly in the docs though.

Reply:
So retrieve the values of the variables as numpy arrays and then set them in the other session.

Reply:
Done. Sorry I had a couple of exams this week, and got to it now.